### DISCUSSION TITLE
What is PAX table access method?

### DISCUSSION THREAD

**Discussion by reshke** [apache/cloudberry#610] [Q&A] [answered]
Im currently working on cherry-picking some GPDB commits related to GPDB7 VACUUM features. This is needed to make [yezzey](https://github.com/yezzey-gp/yezzey)-related patches rebase work easier. 
So, i noticed commit b413e0bce9bffafffa3324fd550a320831a915d5. I wonder what PAX access method is. Is it some UNION store-related tam?

**Comment by jianlirong** [comment, marked as answer]
PAX is a hybrid row-column storage for Cloudberry Database, having the same writer performance as AO tables and the same read performance as AOCS tables. We also have integrated the latest compression algorithms and encoding algorithms (such as dictionary encoding) into the PAX storage. 

We are working on code refactoring and cleaning before making it open-source. Stay tuned!

**Reply by reshke** [reply]
Oh, so PAX actually stands for Partition Attributes Across!
Does your implementation design come from some concrete paper, like AQUMV feature?
https://www.vldb.org/conf/2001/P169.pdf maybe?

**Reply by jianlirong** [reply]
Yes. Actually, many database storage formats, including Snowflake's Micro Partition and Apache ORC, have been influenced by the PAX design philosophy. Our storage format is no exception, although we have placed a greater emphasis on compatibility with PostgreSQL/CBDB data types and the optimization of the CBDB execution engine.

**Comment by jianlirong** [comment]
Beside the performance improvements mentioned above, compared to AO/AOCS tables, we offload the metadata associated with segment files of AO/AOCS tables, say, aoseg and visibility map, to some meta-files stored alongside with the segment files. That makes the separation of compute and storage evolution much easier. 

@reshke , if you are working on migrating [yezzey](https://github.com/yezzey-gp/yezzey) from GPDB to CBDB, personally, I strongly recommend that you should try PAX rather than the AO/AOCS tables.
